internal/tasks/account_management.go

```diff
@@ -0,0 +1,166 @@
+/******************************************************************************
+*
+*  Copyright 2024 SAP SE
+*
+*  Licensed under the Apache License, Version 2.0 (the "License");
+*  you may not use this file except in compliance with the License.
+*  You may obtain a copy of the License at
+*
+*      http://www.apache.org/licenses/LICENSE-2.0
+*
+*  Unless required by applicable law or agreed to in writing, software
+*  distributed under the License is distributed on an "AS IS" BASIS,
+*  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+*  See the License for the specific language governing permissions and
+*  limitations under the License.
+*
+******************************************************************************/
+
+package tasks
+
+import (
+	"context"
+	"database/sql"
+	"encoding/json"
+	"errors"
+	"slices"
+	"time"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sapcc/go-bits/jobloop"
+	"github.com/sapcc/go-bits/logg"
+
+	"github.com/sapcc/keppel/internal/auth"
+	peerclient "github.com/sapcc/keppel/internal/client/peer"
+	"github.com/sapcc/keppel/internal/keppel"
+	"github.com/sapcc/keppel/internal/models"
+)
+
+// EnforceManagedAccounts is a job. Each task creates newly discovered accounts from the driver.
+func (j *Janitor) EnforceManagedAccounts(registerer prometheus.Registerer) jobloop.Job {
+	return (&jobloop.ProducerConsumerJob[[]string]{
+		Metadata: jobloop.JobMetadata{
+			ReadableName: "create new managed accounts",
+			CounterOpts: prometheus.CounterOpts{
+				Name: "keppel_managed_account_creations",
+				Help: "Counter for managed account creations.",
+			},
+		},
+		DiscoverTask: j.discoverManagedAccounts,
+		ProcessTask:  j.enforceManagedAccounts,
+	}).Setup(registerer)
+}
+
+func (j *Janitor) discoverManagedAccounts(_ context.Context, _ prometheus.Labels) (accountNames []string, err error) {
+	managedAccountNames, err := j.amd.ManagedAccountNames()
+	if err != nil {
+		return nil, err
+	}
+
+	var existingAccountNames []string
+	_, err = j.db.Select(&existingAccountNames, "SELECT name FROM accounts WHERE is_managed = true")
+	if err != nil {
+		return nil, err
+	}
+
+	for _, accountName := range managedAccountNames {
+		if !slices.Contains(existingAccountNames, accountName) {
+			accountNames = append(accountNames, accountName)
+		}
+	}
+
+	var toEnforceAccountNames []string
+	_, err = j.db.Select(&toEnforceAccountNames, "SELECT name FROM accounts WHERE is_managed = true and next_account_enforcement_at < $1", j.timeNow())
+	if err != nil {
+		return nil, err
+	}
+	accountNames = append(accountNames, toEnforceAccountNames...)
+
+	if len(accountNames) == 0 {
+		return nil, sql.ErrNoRows
+	}
+
+	return accountNames, nil
```

Comment:
I typed out several review comments that I ended up deleting when I gradually understood more about what this function is doing. This definitely needs some "coding in paragraphs" level comments to outline the overall approach.

Also, I would suggest returning only a single account name at a time to help clarify the approach:

- If there are ManagedAccountNames that do not have a DB record with `is_managed`, return the first such name.
- Otherwise, return the next account name that has `next_account_enforcement_at < NOW()` (in chronological order).